### ISSUE TITLE
Q: Slang for shaders

### ISSUE THREAD

**Issue by edubart** [floooh/sokol#1162]
Since announcement from Khronos about hosting the [slang](https://shader-slang.com/) shading language last month, I wondered what if Sokol supported it to create shaders, it's [playground](https://shader-slang.com/slang-playground/) seems cool. Seems like it could fit as a more modern, more capable, and highly cross platform shading language since it can target all backends Sokol cares, maybe it would simplify `sokol-shdc`.

What are your opinions on the subject?

**Comment by floooh** [comment]
I'm thinking about it but there are a couple of issues to solve:

- how to add the (more or less) engine-specific binding annotations, all Slang code I've seen so far seems to depend on not defining explicit binding information, but let Slang assign the bindings info and then obtain that via reflection
- how to integrate with the sokol-gfx specific code generation - it would be ideal to use Slang as library and build our own cmdline tool around it
- ...but for that it really depends on how big the resulting executable is, and things like whether Slang can be completely statically linked or requires DLLs (the .gitmodules of the Slang project looks kinda scary: https://github.com/shader-slang/slang/blob/master/.gitmodules)

**Comment by edubart** [comment]
> on how big the resulting executable is

Seems like they offer prebuilt binaries for most platforms at releases page, but they have up to 276MB in size, so it's too big.

> Slang can be completely statically linked

From inspecting their binary package, it looks like with some work it could be statically linked, but it could result in a 700MB binary file, ouch.

> it would be ideal to use Slang as library

Since its binaries are very big, maybe expect the user to have a slang compiler available in the system, and invoke `slangc`  command, instead of using it as a library?

Anyway, I have seen Sokol creating its own shader generator to target different backends, SDL3 seems to be on the same path with its own shader compiler. I just thought slang looks like a promising solution for this common problem that cross platform graphics libraries have.

**Comment by floooh** [comment]
> to have a slang compiler available in the system, and invoke slangc command

Yeah that could be the only realistic option, create a small wrapper cmdline tool which calls the slang executable and then uses its output for the sokol-gfx code generation...

Ideally I'd like to give sokol-gfx users multiple options for authoring their shaders (another one could be WGSL as shader authoring language). Just need to find a way which doesn't add too much maintenance overhead.

**Comment by iperov** [comment]
Honestly sokol-shdc scares me away from using sokol at all.
Why can't I use a pure glsl shader?
It ruins the concept of no dependencies and future-proof library.

**Comment by floooh** [comment]
> Why can't I use a pure glsl shader or spirv binary?

Mainly reason is because there's a lot of reflection information extracted from the shader which you would otherwise need to provide manually.

Btw, you *can* use sokol-gfx without sokol-shdc, but in that case you need to provide backend-specific shader code, see the samples here:

- https://github.com/floooh/sokol-samples/tree/master/d3d11
- https://github.com/floooh/sokol-samples/tree/master/glfw
- https://github.com/floooh/sokol-samples/tree/master/html5
- https://github.com/floooh/sokol-samples/tree/master/metal

That also means that other shader compilers could be used instead of sokol-shdc to provide those backend-specific shader source/blobs, but ideally those alternative shader compilers would also extract the reflection information that goes into `sg_shader_desc`.

**Comment by ib00** [comment]
Are there any more concrete plans for supporting slang in sokol?

**Comment by floooh** [comment]
No concrete plans so far. The main problem to solve is how to deal with the highly sokol-gfx specific code-generation and extracting the required reflection information (plus a couple of custom features which are currently handled via '@-tags'), Slang also needs to be flexible enought to annotate resources with Sokol-specific information (like bindslots, or the `@image_sample_type` and `@sampler_type` annotations required for WebGPU.

There's 3 options, all of them kinda suck:

1. fork Slang and add the above features, most flexible but also the most work
2. use an unmodified Slang as library and wrap the above features around it, this assumes that the Slang codebase already provides a flexible enough reflection API and also allows to inject custom attributes without modifying the codebase (big if)
3. call an unmodified Slang binary from a much smaller wrapper executable which implements the sokol-specific features, which even puts more assumptions about how flexible Slang is without requiring code modifications.

**Comment by ib00** [comment]
Thanks for looking into this. I agree that none of these options are straightforward or easy and they
would require significant effort. It's too bad that all graphics APIs and "shading languages" are so 
fragmented.

Has anybody tried to use slang-compiled shaders with sokol_gfx (with manually constructed shader_desc)?

**Comment by floooh** [comment]
> Has anybody tried to use slang-compiled shaders with sokol_gfx (with manually constructed shader_desc)?

Not that I know of. An alternative path I want to look into is to allow sokol-shdc to directly consume SPIRV (basically skip the internal compilation step from GLSL to SPIRV), but this will suffer from a somewhat similar problem of trying to exteact metadata from the SPIRV which might no longer be available. One idea is to provide a separate metadata.json which provides this data...

**Comment by THISISAGOODNAME** [comment]
> Has anybody tried to use slang-compiled shaders with sokol_gfx (with manually constructed shader_desc)?

@ib00 I migrated a simple demo from raw shader.

https://github.com/CU-Production/SokolCSTest/blob/slang/rasterizer_dx_slang_offline_comp.cpp
https://github.com/CU-Production/SokolCSTest/blob/slang/rasterizer_dx_slang_online_comp.cpp

Slang runtime library is about 25MiB, sadly.
